fix: update badge component's style and doc

RyanZhiNie opened the pull request:
## 📑 Description
<!-- Add a brief description of the PR -->
The badge component has been updated in [flowbite v1.6.1 release](https://github.com/themesberg/flowbite/releases/tag/v1.6.1), current badge component styles are not the same as flowbite, so I make this PR. You can view the result in below vercel deployment.


## ✅ Checks
<!-- Make sure your PR passes the tests and do check the following fields as needed - -->

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [ ] I have checked the page with https://validator.unl.edu/
- [x] All the tests have passed
- [x] My pull request is based on the latest commit (not the npm version).

<!--

Sync fork from GitHub dropdown and update your local branch. 

```sh
git pull origin main
git checkout your-branch
git rebase main
npm run test
git push
```

Then create a PR from your GitHub repo.

Or if you are using the command line:

```sh
git checkout main
git fetch upstream // I'm assuming you set the upstream
git merge upstream/main
```

Then change to your branch:

```sh
git checkout my-new-feature
git merge main
```

If you may need to resolve merge conficts if your local branch had unique commits. 

Now you are ready to submit your PR.

It’s a good idea to sync from time to
time, so you aren’t left too far behind the parent branch.

-->

RyanZhiNie commented on src/lib/badges/Badge.svelte:
```diff
@@ -1,26 +1,24 @@
 <script lang="ts">
   import classNames from 'classnames';
-  import CloseButton from '../utils/CloseButton.svelte';
```
This CloseButton component's style is hard to modify, so I directly make changes on a <button> element (see beow).

RyanZhiNie commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,70 @@
+    yellow: 'hover:bg-yellow-200',
+    indigo: 'hover:bg-indigo-200',
+    purple: 'hover:bg-purple-200',
+    pink: 'hover:bg-pink-200'
```
The hover styles in dark mode are not set in flowbite's official examples, so I directly delete previous `dark:hover:` styles, maybe they can be re-added after flowbite makes some official `dark:hover:` style suggestions.

RyanZhiNie commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,70 @@
   let badgeClass: string;
   $: badgeClass = classNames(
     baseClass,
-    large ? 'text-sm font-medium' : 'text-xs font-semibold',
```
According to [flowbite](https://flowbite.com/docs/components/badge/#large-badges), large badges do not change font-weight.

RyanZhiNie commented on src/lib/badges/Badge.svelte:
```diff
@@ -1,26 +1,24 @@
 <script lang="ts">
   import classNames from 'classnames';
-  import CloseButton from '../utils/CloseButton.svelte';
```
This CloseButton component's style is hard to modify, so I directly make changes on a element (see beow).

jjagielka replied:
`CloseButton` is supposed to be re-used among different components so I don't like hardcoding it in the badge.

RyanZhiNie commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,70 @@
   };
 
-  const hovers = {
-    blue: 'hover:bg-blue-200 dark:hover:bg-blue-300',
-    dark: 'hover:bg-gray-200 dark:hover:bg-gray-300',
-    red: 'hover:bg-red-200 dark:hover:bg-red-300',
-    green: 'hover:bg-green-200 dark:hover:bg-green-300',
-    yellow: 'hover:bg-yellow-200 dark:hover:bg-yellow-300',
-    indigo: 'hover:bg-indigo-200 dark:hover:bg-indigo-300',
-    purple: 'hover:bg-purple-200 dark:hover:bg-purple-300',
-    pink: 'hover:bg-pink-200 dark:hover:bg-pink-300'
```
The hover styles in dark mode are not set in flowbite's official examples, so I directly delete previous `dark:hover:` styles, maybe they can be re-added after flowbite makes some official `dark:hover:` style suggestions.

jjagielka commented:
Dismissable badges miss the y-padding and seems like there's something wrong with right padding for the close button.

jjagielka commented on src/lib/badges/Badge.svelte:
```diff
@@ -1,26 +1,24 @@
 
   export let color: 'blue' | 'dark' | 'red' | 'green' | 'yellow' | 'indigo' | 'purple' | 'pink' = 'blue';
   export let large: boolean = false;
+  export let bordered: boolean = false;
```
We use `border` prop on the `Frame`, so it'd be good to be consistent and use the name `border`

RyanZhiNie replied:
I use `bordered` here is to keep it consistent with existing `large` and `rounded`, these three words are all adjectives.
@shinokada which one do you prefer?

shinokada replied:
To be consistent with other components, it will be better to keep the `border`. If you change it, it will affect other components and pages.

RyanZhiNie replied:
Okay, got it. 😃

jjagielka commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,70 @@
   {#if dismissable}
-    <CloseButton {color} on:click={handleHide} size={large ? 'sm' : 'xs'} class="ml-1.5 -mr-1.5" />
-  {/if}
+    <button on:click={handleHide} type="button" class={`${closeBtnBaseClass} ${closeBtnColors[color]}`} aria-label="Remove">
```
As stated already, I'm not a fan of hardcoding the button here. I think it'd be better to use a slot:
```svelte
{#if dismissable}
<slot name="close_button">
  <CloseButton {color} on:click={handleHide} size={large ? 'sm' : 'xs'} class="ml-1.5 -mr-1.5" />
</slot>
{/if}
```

jjagielka commented on src/routes/components/badge.md:
```diff
@@ -68,45 +68,88 @@ Use the following badge elements to indicate counts or labels inside or outside
-<Badge color="dark" class="mr-2">
-<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="mr-1 w-3 h-3"><path stroke-linecap="round" stroke-linejoin="round" d="M12 6v6h4.5m4.5 0a9 9 0 11-18 0 9 9 0 0118 0z" /></svg>3 days ago</Badge>
-<Badge large={true}><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="mr-1 w-4 h-4"><path stroke-linecap="round" stroke-linejoin="round" d="M12 6v6h4.5m4.5 0a9 9 0 11-18 0 9 9 0 0118 0z" /></svg>2 minutes ago</Badge>
+<Badge color="dark" bordered={true}>
```
`<Badge color="dark" bordered>`

jjagielka commented on src/routes/components/badge.md:
```diff
@@ -68,45 +68,88 @@ Use the following badge elements to indicate counts or labels inside or outside
+  <svg aria-hidden="true" class="w-3 h-3 mr-1" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" d="M10 18a8 8 0 100-16 8 8 0 000 16zm1-12a1 1 0 10-2 0v4a1 1 0 00.293.707l2.828 2.829a1 1 0 101.415-1.415L11 9.586V6z" clip-rule="evenodd"></path></svg>
+  3 days ago
+</Badge>
+<Badge bordered={true}>
```
<Badge color="dark" bordered>

RyanZhiNie commented:
I've updated my PR. Thanks for your advice. @jjagielka 

> Dismissable badges miss the y-padding and seems like there's something wrong with right padding for the close button.

In fact, these were not bugs, instead just some padding styles that looked not good. I've updated the dismissable badges example part to make those badges look the same as flowbite's example. I think these styles depend on users' preferences, so I don't modify them directly in the component code. 🤔

jjagielka commented:
As stated already, I'm not a fan of hardcoding the close button within `Badge` component. 

`CloseButton` component is intended to be used throughout different cases to keep the consistent look and feel.

I think it'd be better to use a slot:
```svelte
{#if dismissable}
<slot name="close_button">
  <CloseButton {color} on:click={handleHide} size={large ? 'sm' : 'xs'} class="ml-1.5 -mr-1.5" />
</slot>
{/if}
```

RyanZhiNie commented:
Yes, I've taken your suggestion and already changed to use the `<CloseButton>`, you can view the preview deployment at https://flowbite-svelte-update-imdwy0jc4-themesberg.vercel.app/components/badge#Dismissable_badges

jjagielka commented:
I'm sorry. I was mislead by the fact that you have left `closeBtnBaseClass` and `closeBtnColor` variables in the code.

RyanZhiNie commented:
Never mind. 💙 
 `closeBtnBaseClass` and `closeBtnColor` are just kept to cover `CloseButton`'s styles especially its color palette.

shinokada commented:
@jjagielka Are happy with the changes?
Is this ready to be merged?
Let me know.

jjagielka commented:
> Never mind. blue_heart `closeBtnBaseClass` and `closeBtnColor` are just kept to cover `<CloseButton>`'s styles especially its color palette.

Changing all the style of the component is not real re-use. Let's check what are the differences in `CloseButton` component that makes you unable to use `<CloseButton {color} ..` here.

jjagielka commented:
@RyanZhiNie 
In your impl you have introduced `closeBtnSize` while previously the size of the `CloseButton` and padding was adjusted to badge size automatically.
So it was previously:
```svelte
<Badge dismissable >Dismissable</Badge>
<Badge dismissable large color="red">Dismissable</Badge>
```
and now it is:
```svelte
<Badge dismissable class="!px-2 !py-1" closeBtnSize="xs">Default</Badge>
<Badge dismissable large color="dark" class="!px-2 !py-1">Dark</Badge>
```
Obviously less convenient for the users.

jjagielka commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,91 @@
-    colors[color],
-    href && (hovers[color] ?? hovers.blue),
-    rounded ? 'rounded-full p-1' : 'rounded px-2.5 py-0.5',
+    large ? 'text-sm px-2 py-1' : 'text-xs px-2.5 py-0.5',
```
I've rechecked with Flowbite.com and currently they use `px-2.5 py-05` for all badges regardless the size.

jjagielka commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,91 @@
 </script>
 
-<svelte:element this={href ? 'a' : 'span'} {href} {...$$restProps} class={badgeClass} class:hidden>
+<svelte:element this={href ? 'a' : 'span'} {href} {...$$restProps} class={badgeClass} class:transition-opacity={transition} class:duration-300={transition} class:ease-out={transition} class:opacity-0={transition} class:hidden>
```
This is a suggestion not a bug.
Instead of the lengthly:
 ```
class:transition-opacity={transition} class:duration-300={transition} class:ease-out={transition} class:opacity-0={transition}
```
we can use the advantage of using `classNames` and have just one line added to `badgeClass`:
```
transition && "transition-opacity duration-300 ease-out opacity-0",
```

jjagielka commented on src/lib/badges/Badge.svelte:
```diff
@@ -31,36 +29,91 @@
+  
   const handleHide = () => {
-    hidden = !hidden;
+    transition = true;
```
And once we are on new `badge` version we can add new feature of dispatching the signal informing about the dismissal: `dispatch('close')`

jjagielka commented on src/routes/props/Badge.json:
```diff
@@ -1 +1 @@
-{"props":[["color","'blue' | 'dark' | 'red' | 'green' | 'yellow' | 'indigo' | 'purple' | 'pink' ","'blue'"],["large","boolean ","false"],["href","string | undefined ","undefined"],["rounded","boolean ","false"],["index","boolean ","false"],["dismissable","boolean ","false"],["baseClass","string ","'inline-flex items-center justify-center -mb-0.5'"]]}
\ No newline at end of file
+{"props":[["color","'blue' | 'dark' | 'red' | 'green' | 'yellow' | 'indigo' | 'purple' | 'pink' ","'blue'"],["large","boolean ","false"],["border","boolean ","false"],["href","string | undefined ","undefined"],["rounded","boolean ","false"],["index","boolean ","false"],["dismissable","boolean ","false"],["closeBtnSize","'xs' | 'sm' | 'md' ","'sm'"]]}
```
Re-run `npm run gen:props` as `closeBtnSize` is removed in the latest version, isn't it?

jjagielka reviewed (changes requested):
I'm sorry. Few minor tweaks requested. Then we are good to go.

jjagielka commented:
@shinokada - ready for merge

RyanZhiNie commented:
Please wait a minute, I'll add a small example to show the close event. 🤣

shinokada commented:
Thank you for your contribution. I will publish a new version shortly.
